File: gpu.py
import pynvml

pynvml.nvmlInit()

DEVICE_INDEX = 0  # your 3080
_handle = pynvml.nvmlDeviceGetHandleByIndex(DEVICE_INDEX)


def free_vram_mb() -> int:
    info = pynvml.nvmlDeviceGetMemoryInfo(_handle)
    return info.free // (1024 * 1024)


def total_vram_mb() -> int:
    info = pynvml.nvmlDeviceGetMemoryInfo(_handle)
    return info.total // (1024 * 1024)


def gpu_utilization() -> int:
    rates = pynvml.nvmlDeviceGetUtilizationRates(_handle)
    return rates.gpu  # percentage 0-100

File: gpu.c
#include <nvml.h>
#include <stdint.h>

static nvmlDevice_t device;

void gpu_init(void) {
    nvmlInit();
    nvmlDeviceGetHandleByIndex(0, &device);
}

void gpu_shutdown(void) {
    nvmlShutdown();
}

uint64_t free_vram_mb(void) {
    nvmlMemory_t mem;
    nvmlDeviceGetMemoryInfo(device, &mem);
    return mem.free / (1024 * 1024);
}

uint64_t total_vram_mb(void) {
    nvmlMemory_t mem;
    nvmlDeviceGetMemoryInfo(device, &mem);
    return mem.total / (1024 * 1024);
}

unsigned int gpu_utilization(void) {
    nvmlUtilization_t util;
    nvmlDeviceGetUtilizationRates(device, &util);
    return util.gpu;
}
